paddle/operators/pool_op.cc

```diff
@@ -0,0 +1,170 @@
+    for (size_t i = 0; i < ksize.size(); ++i) {
+      output_shape.push_back(outputSize_pool(in_X->dims()[i + 2], ksize[i],
+                                             paddings[i], strides[i]));
+    }
+    out->Resize(framework::make_ddim(output_shape));
+  }
+};
+
+class PoolOpGrad : public framework::OperatorWithKernel {
+ public:
+  using framework::OperatorWithKernel::OperatorWithKernel;
+
+ protected:
+  void InferShape(const framework::InferShapeContext &ctx) const override {
+    auto in = ctx.Input<Tensor>("X");
+    auto d_in = ctx.Output<framework::LoDTensor>(framework::GradVarName("X"));
+    if (d_in) d_in->Resize(in->dims());
+  }
+};
+
+class Pool3dOpMaker : public framework::OpProtoAndCheckerMaker {
+ public:
+  Pool3dOpMaker(framework::OpProto *proto, framework::OpAttrChecker *op_checker)
+      : OpProtoAndCheckerMaker(proto, op_checker) {
+    AddInput(
+        "X",
+        "The input tensor of pooling operator. "
+        "The format of input tensor is NCDHW. Where N is batch size, C is the "
+        "number of channels, D, H and W is the depth, height and width of "
+        "image.");
+    AddOutput("Out",
+              "The output tensor of pooling operator."
+              "The format of output tensor is also NCDHW.");
+
+    AddAttr<std::string>("pooling_type",
+                         "pooling_type of pooling operator.['max' or 'ave']");
+    AddAttr<std::vector<int>>(
+        "ksize", "pooling size(depth, height, width) of pooling operator.");
+    AddAttr<int>("global_pooling", "whether to use the global_pooling.")
+        .SetDefault(0);
+    AddAttr<std::vector<int>>(
+        "strides", "strides(depth, height, width) of pooling operator.")
+        .SetDefault({1, 1, 1});
+    AddAttr<std::vector<int>>(
+        "paddings", "paddings(depth, height, width) of pooling operator.")
+        .SetDefault({0, 0, 0});
+    AddComment(R"DOC(
+The pooling3d operation calculates the output based on
+the input, pooling_type and ksize, strides, paddings parameters.
+)DOC");
+  }
+};
+
+class Pool2dOpMaker : public framework::OpProtoAndCheckerMaker {
+ public:
+  Pool2dOpMaker(framework::OpProto *proto, framework::OpAttrChecker *op_checker)
+      : OpProtoAndCheckerMaker(proto, op_checker) {
+    AddInput(
+        "X",
+        "The input tensor of pooling operator. "
+        "The format of input tensor is NCHW. Where N is batch size, C is the "
+        "number of channels, H and W is the height and width of image.");
+    AddOutput("Out",
+              "The output tensor of pooling operator."
+              "The format of output tensor is also NCHW.");
+
+    AddAttr<std::string>("pooling_type",
+                         "pooling_type of pooling operator.['max' or 'ave']");
+    AddAttr<std::vector<int>>(
+        "ksize", "pooling size(height, width) of pooling operator.");
+    AddAttr<int>("global_pooling", "whether to use the global_pooling.[0 or 1]")
```

Comment:
"pooling_type" and "ksize"  don't set default values. Should it be judged in InferShape?@hedaoyuan